### ISSUE TITLE
zenoh_ext::QueryingSubscriber optimisation

### ISSUE THREAD

**Issue by JEnoch** [eclipse-zenoh/zenoh#273]
### Describe the feature

In current implementation, when issuing a query, the QueryingSubscriber pushes in a _merge_queue_ the samples resulting of this query and the samples coming directly from the publishers. Then, at query end, the _merge_queue_ is sorted and dedup-ed.

I would be more efficient to use an enhanced _merge_queue_ that sort+dedup on the fly, for each pushed sample.

**Comment by milyin** [comment]
Made request https://github.com/eclipse-zenoh/zenoh/pull/286 with fix for this issue

**Comment by milyin** [comment]
Made updated request https://github.com/eclipse-zenoh/zenoh/issues/273

**Comment by JEnoch** [comment]
Thanks for your contribution @milyin !
I just updated it on 2 aspects:

1. as you might have seen in api-changes branch, the call to `sample.ensure_timestamp()` was removed. The problem calling this operation could be that an historical sample without timestamp (i.e. a reply from the query) could be timestamped with a more recent timestamp than a "live data" (i.e. published by a Publisher during the query).
Therefore I updated your MergeQueue to have a separate Vector for non-timestamped samples. And non-timestamped samples will always be returned in reception order, before any timestamped sample. 
`sample.ensure_timestamp()` is now called only for samples coming from the subscription callback and before insertion into the MergeQueue. Thus, a non-timestamped "live data" cannot be sorted before a timestamped "historical data".

2. I improved the `ensure_timestamp()` operation to return a `&Timestamp` , avoiding a copy.
